Add option to control whether tagstack is copied to new window

yugr opened the pull request:
This is a simple fix for an old issue #5742.

Some people find that default Vim's behavior of copying entire tag stack when splitting a window with unrelated file is confusing. This PR adds a flag (`copytagstack`) to give users control over this.

Please note that, not being a Vim expert, I may miss some hidden issues. I'd be happy to improve/rewrite PR as necessary.

yugr committed:
adds option to control whether tagstack is copied to new window

fixes: #5742

yegappan commented:
> I have not much of an opinion on it, other what I mentioned below.
> @yegappan I think you might have some opinion here?

I'm used to the current behavior where the tag stack is copied to new windows
or tab pages. This allows me to jump to a tag  in a new window while still
retaining the original context in the previous window - helpful for traversing
the tag stack without losing my place.

While adding yet another option to Vim always warrants consideration, I don't
see any major issues with supporting this behavior for users who prefer not to
copy the tag stack. That said, this change may introduce some usability
friction, as users would now need to be aware of which window holds a
particular tag stack in order to traverse it effectively

benknoble commented:
> > > This is a simple fix for an old issue #5742.
> 
> > > Some people find that default Vim's behavior of copying entire tag stack when splitting a window with unrelated file is confusing. This PR adds a flag (`copytagstack`) to give users control over this.
> 
> > Instead of an option, could this be solved at the user-config level with an autocommand that clears the tagstack when a new window is opened?
> 
> You mean something like
> 
> ```
> autocmd BufWinEnter * :some_new_command_to_clean_tag_stack
> ```
> 
> ? I guess this would work too and command to clean tag stack seems useful in general.

Yes, although for some reason I thought there was a function API for clearing it because I remembered working with related stuff in https://github.com/benknoble/tag-tracks

> > Heck, I’m not sure if any plugin authors rely on this behavior.
> 
> True, I'm not sure how to check this...

Me neither at the moment.

benknoble commented:
> > > > This is a simple fix for an old issue #5742.
> > > >
> > > > Some people find that default Vim's behavior of copying entire tag stack when splitting a window with unrelated file is confusing. This PR adds a flag (`copytagstack`) to give users control over this.
> > >
> > > Instead of an option, could this be solved at the user-config level with an autocommand that clears the tagstack when a new window is opened?
> > 
> > You mean something like
> > 
> > ```
> > autocmd BufWinEnter * :some_new_command_to_clean_tag_stack
> > ```
> > 
> > ? I guess this would work too and command to clean tag stack seems useful in general.

Something I omitted from my first reply (below), you probably want the `WinNew` event based on the issue description. 

> Yes, although for some reason I thought there was a function API for clearing it because I remembered working with related stuff in https://github.com/benknoble/tag-tracks

yugr committed:
Address review comments.

yugr committed:
Added a cleartagstack function to clear window tagstack.

This could be used in autocommands like
    autocmd WinNew * :call cleartagstack()
to auto-clear tagstacks of new windows.

yugr commented:
I tried to address all review comments from @chrisbra. I've also added an example `cleartagstack` command to clear tagstack in autocmds as suggested by @benknoble.

The final PR, I assume, will contain either option or command, not both. Using autocmd is fine with me, I'd appreciate if maintainers could comment whether this may be somehow inferior to a full-blown option.

And I still need to figure out the crash for `linux (huge, gcc, uchar, testgui, true, dynamic)` config.

benknoble commented:
> Indeed, it makes sense for `:split` to inherit all source window information, including the tagstack, but there may be cases where you don't want to carry over the tagstack when using `:new`. Since the patch in this PR applies the same behavior to both `:split` and `:new`, it doesn't meet my requirements.  
> 
> By the way, even with the current implementation, you can clear the tagstack only for `:new` by setting up the following auto-command:
> 
> ```vim
> :au BufWinEnter * call settagstack(0, {'items' : []})
> ```
> 
> (In my case I’d use `BufWinEnter`, but for you it might be `BufNew`)

@h-east FWIW I suggested `WinNew` in a previous comment. Though I now wonder in which window the autocommand callback runs.

benknoble commented:
It may just be that I'm on mobile, but the indentation in C files from the cleartagstack patch looks odd.

yugr commented:
> > Indeed, it makes sense for `:split` to inherit all source window information, including the tagstack, but there may be cases where you don't want to carry over the tagstack when using `:new`. Since the patch in this PR applies the same behavior to both `:split` and `:new`, it doesn't meet my requirements.
> > By the way, even with the current implementation, you can clear the tagstack only for `:new` by setting up the following auto-command:
> > ```viml
> > :au BufWinEnter * call settagstack(0, {'items' : []})
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (In my case I’d use `BufWinEnter`, but for you it might be `BufNew`)
> 
> @h-east FWIW I suggested `WinNew` in a previous comment. Though I now wonder in which window the autocommand callback runs.

FYI I observe the following behaviours for different variants of `autocmd ??? :call cleartagstack()`:
1) `WinNew`:
  - `:new`, `:sp`, `:sp some-other-file` - new window has empty tagstack
2) `BufNew`, `BufWinEnter`:
  - `:new`, `:sp some-other-file` - new windows has empty tagstack
  - `:sp`- new window has copy of tagstack
  - in some cases jumping to tag in new file _resets_ existing tagstack (so this is probly unusable)

chrisbra commented:
Can you please try out the autocmd solution? It seems like it is enough and we shouldn't require any code changes here. But I'll leave this open so you can report back how well this approach works in practice.

chrisbra commented:
this has stalled, so closing
